### PR TITLE
Allow any attributes on non-registry overrides

### PR DESCRIPTION
**Pull request by Wyverald**
So this kind of ran away from me...

* The motivator for this PR was to allow any `http_archive` attributes to be specified on `archive_override`. It was kind of a design oversight that we didn't allow this, to be honest; we've slowly accumulated fixes like adding `init_gitsubmodules` to `git_override`, adding `patch_cmds` to `archive_override`, etc., but the better alternative is to just forward all kwargs to the actual underlying repo rule.
* This somehow led to a big refactor where I remove the various `NonRegistryOverride` impls in favor of a simple wrapper around a `RepoSpec`, which led to the creation of a `RepoRuleId` class, which led to a refactor in BzlmodRepoRuleFunction (which also now houses the previous logic to validate labels in `ModuleFileGlobals`).
* Also includes some random Java 21 fixes that I couldn't resist... (`ModuleOverride` is now a sealed interface; added several switch expressions)

The new tests are not passing yet but that'll have to wait until tomorrow.

### CODE REVIEW

**Review by fmeum — APPROVED** [review]
Great refactoring!

**Review by meteorcloudy — APPROVED** [review]
Thanks for fixing this long standing pain point!

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/LocalPathOverride.java [hunk 1]
```diff
@@ -15,38 +15,18 @@
...
- *
- * @param path The path to the local directory where the module contents should be found.
- */
-@AutoCodec
-public record LocalPathOverride(String path) implements NonRegistryOverride {
-  public LocalPathOverride {
-    requireNonNull(path, "path");
-  }
+/** A utility class for {@code local_path_override}. */
+public final class LocalPathOverride {
```
[the remark this answers is not in the document]
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/LocalPathOverride.java]: I essentially wanted something like `ArchiveRepoSpecBuilder` except that the builder would have just a single `setPath` method. Opted to go for `LocalPathRepoSpecs` instead -- the plural usually suggests that it's a utility class (like `Sets`, `Collections`, etc).

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileValue.java [hunk 2]
```diff
@@ -38,7 +38,7 @@ public abstract class BazelLockFileValue implements SkyValue {
   // https://cs.opensource.google/bazel/bazel/+/release-7.3.0:src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java;l=120-127;drc=5f5355b75c7c93fba1e15f6658f308953f4baf51
   // While this hack exists on 7.x, lockfile version increments should be done 2 at a time (i.e.
   // keep this number even).
-  public static final int LOCK_FILE_VERSION = 12;
+  public static final int LOCK_FILE_VERSION = 15;
```
[the remark this answers is not in the document]
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileValue.java]: done

**Review by fmeum — APPROVED** [review]

**Pull request closed without merging**